File: gui_apps/ch02/byteConverter/byteConverterDialog.cpp
// byteConverter/byteConverterDialog.cxx
#include "byteConverterDialog.hxx"
#include <QApplication>
#include <QGridLayout>
#include <QHBoxLayout>
#include <QIntValidator>
#include <QLabel>
#include <QLineEdit>
#include <QPalette>
#include <QPushButton>
#include <QRegExpValidator>
#include <QToolTip>
#include <QVBoxLayout>

// helper function
void setReadOnly(QLineEdit &lineEdit, bool readOnly = true)
{
   QPalette palette = lineEdit.palette();
   QPalette::ColorGroup colorGroup = readOnly ? QPalette::Disabled : QPalette::Active;
   QColor textColor = palette.color(colorGroup, QPalette::WindowText);
   QColor backColor = palette.color(colorGroup, QPalette::Base);
   palette.setColor(QPalette::Base, backColor);
   palette.setColor(QPalette::WindowText, textColor);
   lineEdit.setReadOnly(readOnly);
   lineEdit.setPalette(palette);
}

ByteConverterDialog::ByteConverterDialog()
{
   // set title of dialog
   setWindowTitle(tr("Qt: Byte Converter"));

   // define the labels & edit controls
   QString strPrompt("Type in a decimal number (8 digits max) into\n"
                     "the decimal field to see hex & binary values");
   QLabel *prompt = new QLabel(strPrompt);
   QLabel *decLabel = new QLabel(tr("Decimal:"));
   QLabel *hexLabel = new QLabel(tr("Hex:"));
   QLabel *binLabel = new QLabel(tr("Binary:"));
   decEdit = new QLineEdit;
   hexEdit = new QLineEdit;
   binEdit = new QLineEdit;
   // set their validators
   // QIntValidator *intValidator = new QIntValidator(0, 255, decEdit);
   QRegExp dx("[0-9]{0,8}"); // upto 8 decimal digits 0-9
   QRegExpValidator *intValidator = new QRegExpValidator(dx, decEdit);
   decEdit->setValidator(intValidator);
   QRegExp hx("[0-9A-Fa-f]{1,2}"); // allowe 2 hex chars
   QRegExpValidator *hexValidator = new QRegExpValidator(hx, hexEdit);
   hexEdit->setValidator(hexValidator);
   QRegExp bx("[0-1]{1,8}"); // allow 8 binary chars
   QRegExpValidator *binValidator = new QRegExpValidator(bx, binEdit);
   binEdit->setValidator(binValidator);
   QPushButton *quitBtn = new QPushButton(tr("Quit"));
   quitBtn->setDefault(true);
   quitBtn->setToolTip("Click to Exit");

   // set the hex & bin edits as readOnly
   setReadOnly(*hexEdit);
   setReadOnly(*binEdit);

   // set layout
   // top row - prompt
   QHBoxLayout *promptLayout = new QHBoxLayout();
   promptLayout->addWidget(prompt);
   // middle section: labels + edits in grid layout
   QGridLayout *editLayout = new QGridLayout();
   editLayout->addWidget(decLabel, 0, 0);
   editLayout->addWidget(decEdit, 0, 1);
   editLayout->addWidget(hexLabel, 1, 0);
   editLayout->addWidget(hexEdit, 1, 1);
   editLayout->addWidget(binLabel, 2, 0);
   editLayout->addWidget(binEdit, 2, 1);
   // bottom row - spacer + quitButton
   QHBoxLayout *buttonLayout = new QHBoxLayout();
   buttonLayout->addStretch();
   buttonLayout->addWidget(quitBtn);

   // layout the dialog, with edit grids at the top
   QVBoxLayout *mainLayout = new QVBoxLayout();
   mainLayout->addLayout(promptLayout);
   mainLayout->addLayout(editLayout);
   // mainLayout->addStretch();
   mainLayout->addLayout(buttonLayout);
   setLayout(mainLayout);

   // signals & slots for edits...
   QObject::connect(decEdit, SIGNAL(textChanged(const QString &)), this, SLOT(decChanged(const QString &)));
   // QObject::connect(hexEdit, SIGNAL(textChanged(const QString &)), this, SLOT(hexChanged(const QString&)));
   // QObject::connect(binEdit, SIGNAL(textChanged(const QString &)), this, SLOT(binChanged(const QString&)));

   // clicking quitBtn closes the dialog
   QObject::connect(quitBtn, SIGNAL(clicked()), this, SLOT(accept()));
   // kick off signals & slots
   decEdit->setText("75");
}

void ByteConverterDialog::decChanged(const QString &newValue)
{
   // fires when user enters text in the Decimal edit
   // convert value entered in this text box & update the hex & bin edits
   bool ok;
   int num = newValue.toInt(&ok);

   if (ok) {
      hexEdit->setText("0x" + QString::number(num, 16));
      binEdit->setText("0b" + QString::number(num, 2));
   } else {
      hexEdit->setText("");
      binEdit->setText("");
   }
}

void ByteConverterDialog::hexChanged(const QString &newValue)
{
   bool ok;
   int num = newValue.toInt(&ok, 16);

   if (ok) {
      decEdit->setText(QString::number(num));
      binEdit->setText(QString::number(num, 2));
   } else {
      decEdit->setText("");
      binEdit->setText("");
   }
}

void ByteConverterDialog::binChanged(const QString &newValue)
{
   bool ok;
   int num = newValue.toInt(&ok, 2);

   if (ok) {
      hexEdit->setText(QString::number(num, 16));
      decEdit->setText(QString::number(num));
   } else {
      hexEdit->setText("");
      decEdit->setText("");
   }
}

File: gui_apps/ch02/byteConverter/byteConverterDialog.py
# byteConverterDialog.py: dialog interface, built manually
import sys
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

USING_PYQT6 = True if PYQT_VERSION_STR.startswith('6') else False

p = pathlib.Path(__file__)
# print(f"Current file's path is {str(p)}")
p.parents[2]
# print(f"p.parents[2] = {str(p.parents[2])}")
sys.path.append(str(p.parents[2]))
import globalvars
globalvars.USING_PYQT6 = USING_PYQT6
import utils


class ByteConverterDialog(QDialog):
    def __init__(self, *args, **kwargs):
        super(QDialog, self).__init__(*args, **kwargs)

        # setup Gui
        self.setWindowTitle("PyQt: Byte Converter")

        # build the layout
        # top row - prompt
        self.prompt = QLabel("Type in a decimal number (8 digits max) into\n" +
                             "the decimal field to see hex & binary values")
        # middle - 3 rows of labels + lineedits
        self.decLabel = QLabel("Decimal:")
        self.hexLabel = QLabel("Hex:")
        self.binLabel = QLabel("Binary:")
        self.decEdit = QLineEdit()
        self.hexEdit = QLineEdit()
        self.binEdit = QLineEdit()
        # set validators for line edits
        #self.decValidator = QIntValidator(0, 999, self.decEdit)
        decExpr = QRegExp(r'[0-9]{0,8}')  # just 0-255
        self.decValidator = QRegExpValidator(decExpr, self.decEdit)
        self.decEdit.setValidator(self.decValidator)
        hexExpr = QRegExp(r'[0-9a-fA-F]{1,2}')  # upto 2 chars only!
        self.hexValidator = QRegExpValidator(hexExpr, self.hexEdit)
        self.hexEdit.setValidator(self.hexValidator)
        binExpr = QRegExp(r'[0-1]{1,8}')  # upto 8 chars only!
        self.binValidator = QRegExpValidator(binExpr, self.binEdit)
        self.binEdit.setValidator(self.binValidator)
        # bottom row - hgap - push button
        self.btnQuit = QPushButton("Quit")
        self.btnQuit.setToolTip("Quit Application")

        # to begin with, let's allow editing in ONLY The decEdit
        # self.hexEdit.setReadOnly(True)
        self.setReadOnly(self.hexEdit)
        self.setReadOnly(self.binEdit)

        promptLayout = QHBoxLayout()
        promptLayout.addWidget(self.prompt)

        editLayout = QGridLayout()
        editLayout.addWidget(self.decLabel, 0, 0)
        editLayout.addWidget(self.decEdit, 0, 1)
        editLayout.addWidget(self.hexLabel, 1, 0)
        editLayout.addWidget(self.hexEdit, 1, 1)
        editLayout.addWidget(self.binLabel, 2, 0)
        editLayout.addWidget(self.binEdit, 2, 1)

        btnLayout = QHBoxLayout()
        btnLayout.addStretch()
        btnLayout.addWidget(self.btnQuit)

        mainLayout = QVBoxLayout()
        mainLayout.addLayout(promptLayout)
        mainLayout.addLayout(editLayout)
        mainLayout.addLayout(btnLayout)
        self.setLayout(mainLayout)

        # setup signals & slots
        self.decEdit.textChanged.connect(self.decChanged)
        self.hexEdit.textChanged.connect(self.hexChanged)
        # self.binEdit.textChanged.connect(self.binChanged)
        self.btnQuit.clicked.connect(QApplication.instance().quit)

    def decChanged(self):
        # NOTE: since we have applied a validator to decEdit
        # we can be sure that decEdit holds either a blank string or
        # a valid string representation of an integer
        decValue = self.decEdit.text()
        if len(decValue.strip()) != 0:
            numDecValue = int(decValue, base=10)
            self.hexEdit.setText(hex(numDecValue))
            self.binEdit.setText(bin(numDecValue))
        else:
            self.hexEdit.setText("")
            self.binEdit.setText("")

    def hexChanged(self):
        # NOTE: since we have applied a validator to hexEdit
        # we can be sure that it holds either a blank string or
        # a valid string representation of a hex number
        hexValue = self.hexEdit.text()
        if len(hexValue.strip()) != 0:
            numHexValue = int(hexValue, base=16)
            self.decEdit.setText(f"{numHexValue}")
            self.binEdit.setText(bin(numHexValue))
        else:
            self.decEdit.setText("")
            self.binEdit.setText("")

    def setReadOnly(self, lineEdit, mode=True):
        assert type(lineEdit) == type(
            self.decEdit), "Expecting QLineEdit as arg to setReadOnly"
        # @see: https://stackoverflow.com/questions/23915700/how-to-make-a-qlineedit-not-editable-in-windows
        # we need to set lineEdit to read only & modify the palette as well
        lineEditPalette = lineEdit.palette()
        colorGroup = QPalette.Disabled if mode == True else QPalette.Active
        textColor = lineEditPalette.color(colorGroup, QPalette.WindowText)
        backColor = lineEditPalette.color(colorGroup, QPalette.Base)
        lineEditPalette.setColor(QPalette.Base, backColor)
        lineEditPalette.setColor(QPalette.WindowText, backColor)

        lineEdit.setReadOnly(mode)
        lineEdit.setPalette(lineEditPalette)


def main():
    app = QApplication(sys.argv)
    # font = QFont(app.font().family(), 12)
    # app.setFont(font)
    app.setFont(QApplication.font("QMenu"))

    w = ByteConverterDialog()
    # w.setFont(QApplication.font("QPushButton"))
    w.decEdit.setText("75")
    w.show()
    return app.exec_()


if __name__ == '__main__':
    main()
